File: common/src/test/java/com/hartwig/actin/json/JsonTest.java
package com.hartwig.actin.json;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import java.time.LocalDate;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import org.junit.Test;

public class JsonTest {

    private static final double EPSILON = 1.0E-10;

    @Test
    public void canExtractObjects() {
        JsonObject object = new JsonObject();

        object.add("object", new JsonObject());
        assertNotNull(Json.object(object, "object"));

        object.add("null", null);
        assertNull(Json.nullableObject(object, "null"));
    }

    @Test
    public void canExtractArrays() {
        JsonObject object = new JsonObject();

        object.add("array1", new JsonArray());
        assertNotNull(Json.array(object, "array1"));

        object.addProperty("nullable", (String) null);
        assertNull(Json.nullableStringList(object, "nullable"));

        JsonArray array = new JsonArray();
        array.add("value1");
        array.add("value2");
        object.add("array2", array);
        assertEquals(2, Json.nullableStringList(object, "array2").size());

        object.addProperty("string", "string");
        assertEquals(1, Json.nullableStringList(object, "string").size());
    }

    @Test
    public void canExtractStrings() {
        JsonObject object = new JsonObject();

        object.addProperty("nullable", (String) null);
        assertNull(Json.nullableString(object, "nullable"));

        object.addProperty("string", "value");
        assertEquals("value", Json.nullableString(object, "string"));
    }

    @Test
    public void canExtractNumbers() {
        JsonObject object = new JsonObject();

        object.addProperty("nullable", (String) null);
        assertNull(Json.nullableNumber(object, "nullable"));

        object.addProperty("number", 12.4);
        assertEquals(12.4, Json.nullableNumber(object, "number"), EPSILON);
    }

    @Test
    public void canExtractIntegers() {
        JsonObject object = new JsonObject();

        object.addProperty("nullable", (String) null);
        assertNull(Json.nullableInteger(object, "nullable"));

        object.addProperty("integer", 8);
        assertEquals(8, (int) Json.nullableInteger(object, "integer"));
    }

    @Test
    public void canExtractBooleans() {
        JsonObject object = new JsonObject();

        object.addProperty("nullable", (String) null);
        assertNull(Json.nullableBool(object, "nullable"));

        object.addProperty("bool", true);
        assertTrue(Json.nullableBool(object, "bool"));
    }

    @Test
    public void canExtractDates() {
        JsonObject object = new JsonObject();

        object.addProperty("nullable", (String) null);
        assertNull(Json.nullableDate(object, "nullable"));

        JsonObject dateObject = new JsonObject();
        dateObject.addProperty("year", 2018);
        dateObject.addProperty("month", 4);
        dateObject.addProperty("day", 6);
        object.add("date", dateObject);
        assertEquals(LocalDate.of(2018, 4, 6), Json.nullableDate(object, "date"));
    }
}

File: common/src/test/kotlin/com/hartwig/actin/util/json/JsonTest.kt
package com.hartwig.actin.util.json

import com.fasterxml.jackson.databind.node.JsonNodeFactory
import com.hartwig.actin.util.json.Json.array
import com.hartwig.actin.util.json.Json.bool
import com.hartwig.actin.util.json.Json.date
import com.hartwig.actin.util.json.Json.integer
import com.hartwig.actin.util.json.Json.nullableBool
import com.hartwig.actin.util.json.Json.nullableDate
import com.hartwig.actin.util.json.Json.objectNode
import com.hartwig.actin.util.json.Json.optionalArray
import com.hartwig.actin.util.json.Json.optionalBool
import com.hartwig.actin.util.json.Json.optionalObject
import com.hartwig.actin.util.json.Json.optionalString
import com.hartwig.actin.util.json.Json.optionalStringList
import com.hartwig.actin.util.json.Json.string
import com.hartwig.actin.util.json.Json.stringList
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.time.LocalDate

class JsonTest {

    private val factory = JsonNodeFactory.instance

    @Test
    fun `Should extract objects`() {
        val obj = factory.objectNode()
        assertThat(optionalObject(obj, "object")).isNull()

        obj.set<com.fasterxml.jackson.databind.JsonNode>("object", factory.objectNode())
        assertThat(optionalObject(obj, "object")).isNotNull
        assertThat(objectNode(obj, "object")).isNotNull
    }

    @Test
    fun `Should extract arrays`() {
        val obj = factory.objectNode()
        assertThat(optionalArray(obj, "array1")).isNull()

        obj.set<com.fasterxml.jackson.databind.JsonNode>("array1", factory.arrayNode())
        assertThat(optionalArray(obj, "array1")).isNotNull
        assertThat(array(obj, "array1")).isNotNull
    }

    @Test
    fun `Should extract string lists`() {
        val obj = factory.objectNode()
        obj.putNull("nullable")
        assertThat(optionalStringList(obj, "array1")).isNull()

        val array = factory.arrayNode().add("value1").add("value2")
        obj.set<com.fasterxml.jackson.databind.JsonNode>("array1", array)
        assertThat(optionalStringList(obj, "array1")).hasSize(2)
        assertThat(stringList(obj, "array1")).hasSize(2)
    }

    @Test
    fun `Should treat single string as singleton list`() {
        val obj = factory.objectNode().put("scalar", "only")
        assertThat(stringList(obj, "scalar")).containsExactly("only")
    }

    @Test
    fun `Should extract strings`() {
        val obj = factory.objectNode()
        assertThat(optionalString(obj, "string")).isNull()

        obj.put("string", "value")
        assertThat(optionalString(obj, "string")).isEqualTo("value")
        assertThat(string(obj, "string")).isEqualTo("value")
    }

    @Test
    fun `Should extract integers`() {
        val obj = factory.objectNode().put("integer", 8)
        assertThat(integer(obj, "integer")).isEqualTo(8)
    }

    @Test
    fun `Should extract booleans`() {
        val obj = factory.objectNode()
        assertThat(optionalBool(obj, "bool")).isNull()
        obj.putNull("nullable")
        assertThat(nullableBool(obj, "nullable")).isNull()

        obj.put("bool", true)
        assertThat(nullableBool(obj, "bool")!!).isTrue
        assertThat(bool(obj, "bool")).isTrue
    }

    @Test
    fun `Should extract dates`() {
        val obj = factory.objectNode()
        obj.putNull("nullable")
        assertThat(nullableDate(obj, "nullable")).isNull()

        val dateNode = factory.objectNode().put("year", 2026).put("month", 6).put("day", 8)
        obj.set<com.fasterxml.jackson.databind.JsonNode>("date", dateNode)
        assertThat(date(obj, "date")).isEqualTo(LocalDate.of(2026, 6, 8))
    }
}
